### PR TITLE
[Form] keep valid submitted choices when additional choices are submitted

### LINKED ISSUE #9738: [Form] Keep selected choices when extra choices have been submitted
I have an expanded form type. If I load the page, then someone deletes one of the entities that is in my available choices and I select/submit it I get an error. This behaviour is correct.

However, all the valid options I have chosen have no been unselected and the data reverts to the original value. Is there a way to stop this and use the values that were selected regardless of whether it is valid/invalid. Or potentially even just ignore invalid values?

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #9738
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by xabbuh** [comment]
Status: Needs work

**Comment by xabbuh** [comment]
Status: Needs Review

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 1]
```diff
@@ -27,31 +27,41 @@
...
+    public function __construct(ChoiceListFactoryInterface $choiceListFactory = null, $translator = null)
     {
         $this->choiceListFactory = $choiceListFactory ?: new CachingFactoryDecorator(
             new PropertyAccessDecorator(
                 new DefaultChoiceListFactory()
             )
         );
+
+        if (null !== $translator && !$translator instanceof LegacyTranslatorInterface && !$translator instanceof TranslatorInterface) {
+            throw new \TypeError(sprintf('Argument 1 passed to "%s()" must be an instance of "%s", "%s" given.', __METHOD__, TranslatorInterface::class, \is_object($translator) ? \get_class($translator) : \gettype($translator)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: Argument 2

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 2]
```diff
@@ -97,6 +109,10 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 // Convert the submitted data to a string, if scalar, before
                 // casting it to an array
                 if (!\is_array($data)) {
+                    if ($form->getConfig()->getOption('multiple', false)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: not `$options['multiple']`? (same in other similar places in the PR)

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 3]
```diff
@@ -27,31 +27,41 @@
...
+use Symfony\Component\Translation\TranslatorInterface as LegacyTranslatorInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
 
 class ChoiceType extends AbstractType
 {
     private $choiceListFactory;
+    private $translator;
 
-    public function __construct(ChoiceListFactoryInterface $choiceListFactory = null)
+    public function __construct(ChoiceListFactoryInterface $choiceListFactory = null, $translator = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php:50]: we should add a docblock `@param TranslatorInterface`

**Comment by xabbuh** [comment]
@yceruto Can you have a look here please? :)

**Review by yceruto — APPROVED** [review]
LGTM, thanks!

**Pull request merged**